wal.go

```diff
@@ -95,3 +100,68 @@ func (b *azureSecretBackend) rollbackRootWAL(ctx context.Context, req *logical.R
 
 	return nil
 }
+
+type walAppRoleAssign struct {
+	SpID          string
+	AssignmentIDs []string
+	AzureRoles    []*AzureRole
+	Expiration    time.Time
+}
+
+func (b *azureSecretBackend) rollbackRoleAssignWAL(ctx context.Context, req *logical.Request, data interface{}) error {
+	// Decode the WAL data
+	var entry walAppRoleAssign
+	d, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		DecodeHook: mapstructure.StringToTimeHookFunc(time.RFC3339),
+		Result:     &entry,
+	})
+	if err != nil {
+		return err
+	}
+	err = d.Decode(data)
+	if err != nil {
+		return err
+	}
+
+	client, err := b.getClient(ctx, req.Storage)
+	if err != nil {
+		return err
+	}
+
+	b.Logger().Debug("rolling back role assignments for service principal", "ID", entry.SpID)
+
+	// Return if there aren't any roles to unassign
+	if entry.AzureRoles == nil {
+		b.Logger().Error("no azure roles associated with role")
+		return nil
+	}
+
+	// Assemble all App Role Assignment IDs
+	var roleAssignments []string
+	for i, assignmentID := range entry.AssignmentIDs {
+		if entry.AzureRoles[i] == nil {
+			return fmt.Errorf("azure role was nil")
+		}
+		roleAssignments = append(roleAssignments, fmt.Sprintf("%s/providers/Microsoft.Authorization/roleAssignments/%s",
+			entry.AzureRoles[i].Scope,
+			assignmentID))
+	}
+
+	// Check any errors to filter out expected responses. Azure will return
+	// a 204 when trying to delete a role assignment that has already been
+	// deleted, or does not exist. We may hit this case during rollback.
+	if err := client.unassignRoles(ctx, roleAssignments); err != nil {
+		for _, e := range err.(*multierror.Error).Errors {
+			switch {
+			case strings.Contains(e.Error(), "StatusCode=204"):
+				b.Logger().Trace("role assignment already deleted or does not exist", "err", e.Error())
+			default:
+				return fmt.Errorf("rollback error unassinging role: %w", e)
+			}
+		}
+		if time.Now().After(entry.Expiration) {
+			return nil
```

Comment:
I'm wondering if we should log something here to indicate that the WAL entry expired and there may be resources that aren't cleaned up 🤔 Looks like we don't above in `rollbackAppWAL`, but it might be useful to know when this happened.